sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/directconnectivity/rntbd/ProactiveOpenConnectionsProcessor.java

```diff
@@ -0,0 +1,195 @@
+    private final AtomicReference<AsyncDocumentClient.OpenConnectionAggressivenessHint> aggressivenessHint;
+    private static final Map<AsyncDocumentClient.OpenConnectionAggressivenessHint, ConcurrencyConfiguration> concurrencySettings = new HashMap<>();
+
+    static {
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.DEFENSIVE, new ConcurrencyConfiguration(1, 1));
+        concurrencySettings.put(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE, new ConcurrencyConfiguration(Configs.getCPUCnt(), Configs.getCPUCnt()));
+    }
+
+    public ProactiveOpenConnectionsProcessor() {
+        this.openConnectionsTaskSink = Sinks.many().multicast().onBackpressureBuffer();
+        this.openConnectionsTaskSinkBackUp = Sinks.many().multicast().onBackpressureBuffer();
+        this.aggressivenessHint = new AtomicReference<>(AsyncDocumentClient.OpenConnectionAggressivenessHint.AGGRESSIVE);
+        this.totalEstablishedConnections = new AtomicInteger(0);
+        this.endpointToMinConnections = new ConcurrentHashMap<>();
+    }
+
+    public synchronized void submitOpenConnectionTask(OpenConnectionOperation openConnectionOperation) {
+
+        String addressUriAsString = openConnectionOperation.getAddressUri().getURIAsString();
+        boolean isEndpointTaskInSink = endpointToMinConnections.putIfAbsent(addressUriAsString, openConnectionOperation.getMinConnectionsRequiredForEndpoint()) != null;
+
+        endpointToMinConnections.computeIfPresent(addressUriAsString, (s, min) -> {
+            openConnectionOperation.setMinConnectionsRequiredForEndpoint(Math.max(openConnectionOperation.getMinConnectionsRequiredForEndpoint(), min));
+            return Math.max(openConnectionOperation.getMinConnectionsRequiredForEndpoint(), min);
+        });
+
+        if (!isEndpointTaskInSink) {
+            openConnectionsTaskSink.tryEmitNext(openConnectionOperation);
+            openConnectionsTaskSinkBackUp.tryEmitNext(openConnectionOperation);
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        completeSink(openConnectionsTaskSink);
+        completeSink(openConnectionsTaskSinkBackUp);
+    }
+
+    public synchronized ParallelFlux<OpenConnectionResponse> getOpenConnectionsPublisher() {
+
+        ConcurrencyConfiguration concurrencyConfiguration = concurrencySettings.get(aggressivenessHint.get());
+
+        return Flux.from(openConnectionsTaskSink.asFlux())
+                .publishOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                .parallel(concurrencyConfiguration.openConnectionOperationEmissionConcurrency)
+                .runOn(CosmosSchedulers.OPEN_CONNECTIONS_BOUNDED_ELASTIC)
+                .flatMap(openConnectionOperation -> {
+                    if (totalEstablishedConnections.get() < MaxConnectionsToOpenAcrossAllEndpoints) {
+
+                        endpointToMinConnections.remove(openConnectionOperation.getAddressUri().getURIAsString());
```

Comment:
should we remove only after for that endpoint, SDK has opened enough connection?

Reply:
It can be removed. Then `IOpenConnectionsHandler` will decide whether to re-submit the open connection task for the endpoint to the sink and the process repeats.